<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
!pip install psycopg2

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [2]:
import pandas as pd
import psycopg2

In [3]:
# Параметры подключения из юнита 1. Работа с базой данных из Python
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [4]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [5]:
# текст запроса
query_3_1 = f'''
    SELECT
        COUNT(id)
    FROM 
        public.vacancies
'''

In [6]:
# результат запроса
df1 = pd.read_sql_query(query_3_1, connection)
display(df1)

/usr/local/lib/python3.9/dist-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [7]:
# текст запроса
query_3_2 = f'''
    SELECT
        COUNT(id)
    FROM
        public.employers    
'''

In [8]:
# результат запроса
df2 = pd.read_sql_query(query_3_2, connection)
display(df2)

/usr/local/lib/python3.9/dist-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [9]:
# текст запроса
query_3_3 = f'''
    SELECT
        COUNT(id)
    FROM
        public.areas          
'''

In [10]:
# результат запроса
df3 = pd.read_sql_query(query_3_3, connection)
display(df3)

/usr/local/lib/python3.9/dist-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [11]:
# текст запроса
query_3_4 = f'''
    SELECT
        COUNT(id)
    FROM
        public.industries        
'''

In [12]:
# результат запроса
df4 = pd.read_sql_query(query_3_4, connection)
display(df4)

/usr/local/lib/python3.9/dist-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0    294

***

# Выводы по предварительному анализу данных
В результате предварительного анализа данных мы выяснили, что в нашей базе имеется: 
<br>Вакансий - 49197, 
<br>Работодателей - 23501, 
<br>Регионов - 1362, 
<br>Сфер деятельности - 294 

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [13]:
# текст запроса
query_4_1 = f'''
    SELECT
        COUNT(v.id) ctn,
        a.name area

    FROM 
        public.vacancies v
    JOIN
        public.areas a ON v.area_id = a.id

    GROUP BY a.name
    ORDER BY ctn DESC
'''

In [14]:
# результат запроса
df5 = pd.read_sql_query(query_4_1, connection)
display(df5)

/usr/local/lib/python3.9/dist-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


ctn             area
0    5333           Москва
1    2851  Санкт-Петербург
2    2112            Минск
3    2006      Новосибирск
4    1892           Алматы
..    ...              ...
764     1       Тарко-Сале
765     1    Новоаннинский
766     1            Бирск
767     1           Сасово
768     1             Анна

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [15]:
# текст запроса
query_4_2 = f'''
    SELECT
        COUNT(v.id)
    FROM
        public.vacancies v
   
    WHERE v.salary_from IS NOT NULL
        OR v.salary_to IS NOT NULL   
'''

In [16]:
# результат запроса
df6 = pd.read_sql_query(query_4_2, connection)
display(df6)

/usr/local/lib/python3.9/dist-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [17]:
# текст запроса
query_4_3 = f'''
    SELECT
        ROUND(AVG(v.salary_from)) avr_salary_from,
        ROUND(AVG(v.salary_to)) avr_salary_to
    FROM
        public.vacancies v    
'''

In [18]:
# результат запроса
df7 = pd.read_sql_query(query_4_3, connection)
display(df7)

/usr/local/lib/python3.9/dist-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


avr_salary_from  avr_salary_to
0          71065.0       110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [19]:
# текст запроса
query_4_4 = f'''
    SELECT
        v.schedule,
        v.employment,
        count(v.id)
    FROM
        public.vacancies v

    GROUP BY 1, 2 
    ORDER BY 3 DESC       
'''

In [20]:
# результат запроса
df8 = pd.read_sql_query(query_4_4, connection)
display(df8)

/usr/local/lib/python3.9/dist-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [21]:
# текст запроса
query_4_5 = f'''
    SELECT
        v.experience,
        COUNT(v.id)
    FROM
        public.vacancies v

    GROUP BY 1
    ORDER BY 2      
'''

In [22]:
# результат запроса
df9 = pd.read_sql_query(query_4_5, connection)
display(df9)

/usr/local/lib/python3.9/dist-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

# Выводы по детальному анализу вакансий
В результате детального анализа вакансий выяснили:
<br>1) Самое большое количество кандидатов как и ожидалось сосредоточенно в крупных городах: Москва, Санкт-Петербург. При этом в Москва является самым активным рынком труда.
<br>2) Почти в половине (24073) вакансий заполненно хотя бы одно из двух полей с зарплатой. Таким образом, в больше чем половине вакансий не заполнены поля с указанием зарплаты.
<br>3) Среднее значение для нижней границы зарплатной вилки - 71065 руб.
<br>  Среднее значение для верхней границы зарплатной вилки - 110537 руб.
<br>4) Наиболее популярное сочетание рабочего графика и типа трудоустройства "Полный день - Полная занятость"
<br>5) Наиболее встречаемый опыт работы от 1 года до 3 лет. Таким образом, сервисом hh.ru в основном пользуются претенденты с минимальным опытом работы. 


# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [23]:
# текст запроса
query_5_1 = f'''
    SELECT
        e.name employer_name,
        COUNT(v.name) ctn
    FROM
        public.employers e
    JOIN 
        public.vacancies v ON e.id = v.employer_id

    GROUP BY 1
    ORDER BY 2 DESC
    LIMIT 10            
'''

In [24]:
# результат запроса
df10 = pd.read_sql_query(query_5_1, connection)
display(df10)

/usr/local/lib/python3.9/dist-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


employer_name   ctn
0                 Яндекс  1933
1             Ростелеком   491
2               Тинькофф   444
3                   СБЕР   428
4          Газпром нефть   331
5            ИК СИБИНТЕК   327
6                    МТС   292
7                DataArt   247
8  Совкомбанк Технологии   204
9             Первый Бит   176

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [25]:
# текст запроса
query_5_2 = f'''
    SELECT
        a.name area,
        COUNT(v.name) qtn_vacancy,
        COUNT(e.name) qtn_employer
    FROM 
        public.areas a
    LEFT JOIN
        public.vacancies v ON v.area_id = a.id
    LEFT JOIN
        public.employers e ON e.area = a.id 
   
    WHERE v.name IS NULL
    GROUP BY 1
    ORDER BY 3 DESC
'''

In [26]:
# результат запроса
df11 = pd.read_sql_query(query_5_2, connection)
display(df11)

/usr/local/lib/python3.9/dist-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


area  qtn_vacancy  qtn_employer
0                            Россия            0           410
1                         Казахстан            0           207
2                Московская область            0            75
3                Краснодарский край            0            19
4                          Беларусь            0            18
..                              ...          ...           ...
588                        Кувандык            0             0
589                        Куйбышев            0             0
590  Куйбышево (Ростовская область)            0             0
591                        Култаево            0             0
592                        Кульсары            0             0

[593 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [27]:
# текст запроса
query_5_3 = f'''
    SELECT
        e.name employer,
        COUNT(DISTINCT a.name) qtn_areas
    FROM
        public.vacancies v
    JOIN   
        public.employers e ON e.id = v.employer_id
    JOIN  
        public.areas a ON a.id = v.area_id

    GROUP BY 1
    ORDER BY 2 DESC       
'''

In [28]:
# результат запроса
df12 = pd.read_sql_query(query_5_3, connection)
display(df12)

/usr/local/lib/python3.9/dist-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


employer  qtn_areas
0                      Яндекс        181
1                  Ростелеком        152
2                  Спецремонт        116
3      Поляков Денис Иванович         88
4                   ООО ЕФИН          71
...                       ...        ...
14761                  UniSol          1
14762            UNISTORY LLC          1
14763                   UNIT6          1
14764     United Distribution          1
14765              UNITEL LLC          1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [29]:
# текст запроса
query_5_4 = f'''
    SELECT 
        COUNT(e.id) qtn
    FROM
        public.employers e
    LEFT JOIN
        public.employers_industries ei ON e.id = ei.employer_id
    LEFT JOIN             
        public.industries i ON ei.industry_id = i.id

    WHERE i.name IS NULL
'''

In [30]:
# результат запроса
df13 = pd.read_sql_query(query_5_4, connection)
df13

/usr/local/lib/python3.9/dist-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


qtn
0  8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [31]:
# текст запроса
query_5_5 = f'''
    SELECT
        e.name,
        COUNT(ei.industry_id) qtn
    FROM
        public.employers e
    JOIN
        public.employers_industries ei ON e.id = ei.employer_id

    GROUP BY 1
    HAVING COUNT(ei.industry_id) = 4
    ORDER BY 1
'''

In [32]:
# результат запроса
df14 = pd.read_sql_query(query_5_5, connection)
df14

/usr/local/lib/python3.9/dist-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name  qtn
0        101 Интернет    4
1            21vek.by    4
2                2ГИС    4
3                  2К    4
4         4 пикселя +    4
...               ...  ...
1131            ЮРИОН    4
1132  ЮТИП Технологии    4
1133           ЯКласс    4
1134            ЯрНео    4
1135    Яшин&Партнёры    4

[1136 rows x 2 columns]

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [33]:
# текст запроса
query_5_6 = f'''
    SELECT
        COUNT(e.id) qtn
    FROM
        public.employers e
    LEFT JOIN
        public.employers_industries ei ON e.id = ei.employer_id
    LEFT JOIN             
        public.industries i ON ei.industry_id = i.id

    WHERE i.name = 'Разработка программного обеспечения'        
'''

In [34]:
# результат запроса
df15 = pd.read_sql_query(query_5_6, connection)
df15

/usr/local/lib/python3.9/dist-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


qtn
0  3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [35]:
!pip install requests 

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [36]:
!pip install beautifulsoup4

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [37]:
# код для получения списка городов-милионников
import requests
from bs4 import BeautifulSoup
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
response = requests.get(url)
page = BeautifulSoup(response.text, 'html.parser')
print(response)
print('------')
links = page.find_all('a')
cities_millionaire = tuple(link.text for link in links[45:62])
print(cities_millionaire)

<Response [200]>
------
('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Челябинск', 'Красноярск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Омск', 'Краснодар', 'Воронеж', '[a]', 'Пермь', 'Волгоград')


In [38]:
# текст запроса
query_5_7 = f'''
    SELECT
        a.name,
        COUNT(v.name) qtn
    FROM
        public.vacancies v
    JOIN
        public.employers e ON v.employer_id = e.id AND e.name = 'Яндекс'        
    JOIN
        public.areas a ON v.area_id = a.id
    WHERE a.name IN {cities_millionaire}
    GROUP BY 1

    UNION ALL
    
    SELECT
        'Total',
        COUNT(v.name)
    FROM
        public.vacancies v
    JOIN
        public.employers e ON v.employer_id = e.id AND e.name = 'Яндекс'        
    JOIN
        public.areas a ON v.area_id = a.id
    WHERE a.name IN {cities_millionaire}
    ORDER BY 2             
'''

In [39]:
# результат запроса
df16 = pd.read_sql_query(query_5_7, connection)
df16

/usr/local/lib/python3.9/dist-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name  qtn
0              Омск   21
1         Челябинск   22
2        Красноярск   23
3         Волгоград   24
4             Пермь   25
5            Казань   25
6    Ростов-на-Дону   25
7               Уфа   26
8            Самара   26
9         Краснодар   30
10          Воронеж   32
11      Новосибирск   35
12  Нижний Новгород   36
13     Екатеринбург   39
14  Санкт-Петербург   42
15           Москва   54
16            Total  485

***

# Выводы по анализу работодателей
В первой пятерке работодателей по количеству вакансий находятся такие компании как: Яндекс, Ростелеком, Тинькофф, СБЕР, Газпром нефть. При этом вакансии Яндекса опубликованны в 181 регионе, Ростелекома - в 152 регионе. 
<br>У 8419 работодателей не указана сфера деятельности. это ~36% от общего числа работодателей.  
<br>У 3553 работодателей в качестве сферы деятельности указана Разработка программного обеспечения. Это ~15% от общего числа работодателей. 
<br>Вакансии компании Яндекс представленны во всех городах миллиониках, больше всего в Москве.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [40]:
# текст запроса
query_6_1 = f'''
    SELECT
        COUNT(v.name)
    FROM
        public.vacancies v

    WHERE name ILIKE '%data%' OR name ILIKE '%данн%'  
'''

In [41]:
# результат запроса
df17 = pd.read_sql_query(query_6_1, connection)
df17

/usr/local/lib/python3.9/dist-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [42]:
# текст запроса
query_6_2 = f'''
    SELECT
        COUNT(v.id)
    FROM
        public.vacancies v

    WHERE 
        (name ILIKE '%data scientist%'
        OR name ILIKE '%data science%'
        OR name ILIKE '%исследователь данных%'
        OR (name LIKE '%ML%' AND name NOT ILIKE '%HTML%')
        OR name ILIKE '%machine learning%'
        OR name ILIKE '%машинн%обучен%')
        
        AND 

        (name ILIKE '%junior%'
        OR experience = 'Нет опыта'
        OR employment = 'Стажировка')     
'''

In [43]:
# результат запроса
df19 = pd.read_sql_query(query_6_2, connection)
df19

/usr/local/lib/python3.9/dist-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [44]:
# текст запроса
query_6_3 = f'''
    SELECT
        COUNT(v.id)
    FROM
        public.vacancies v

    WHERE 
        (name ILIKE '%data scientist%'
        OR name ILIKE '%data science%'
        OR name ILIKE '%исследователь данных%'
        OR (name LIKE '%ML%' AND name NOT ILIKE '%HTML%')
        OR name ILIKE '%machine learning%'
        OR name ILIKE '%машинн%обучен%')

        AND

        (key_skills ILIKE '%SQL%'
        OR key_skills ILIKE '%postgres%')   
'''        

In [45]:
# результат запроса
df20 = pd.read_sql_query(query_6_3, connection)
df20

/usr/local/lib/python3.9/dist-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0    201

In [46]:
query_test = f'''
    SELECT
        v.key_skills
    FROM
        public.vacancies v  
'''

df_test = pd.read_sql_query(query_test, connection)
df_test['key_skills'][0]

/usr/local/lib/python3.9/dist-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


'Пользователь ПК\tРабота в команде\tРемонт ноутбуков\tДиагностика\tОпытный пользователь ПК\tРемонт\tУстановка ОС\tРемонт компьютеров\tНавыки продаж'

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [47]:
# текст запроса
query_6_4 = f'''
    SELECT
        COUNT(v.id)
    FROM
        public.vacancies v

    WHERE 
        (name ILIKE '%data scientist%'
        OR name ILIKE '%data science%'
        OR name ILIKE '%исследователь данных%'
        OR (name LIKE '%ML%' AND name NOT ILIKE '%HTML%')
        OR name ILIKE '%machine learning%'
        OR name ILIKE '%машинн%обучен%')

        AND

        key_skills ILIKE '%Python%'   
'''       

In [48]:
# результат запроса
df21 = pd.read_sql_query(query_6_4, connection)
df21

/usr/local/lib/python3.9/dist-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [49]:
# текст запроса
query_6_5 = f'''
    SELECT
        (LENGTH(v.key_skills) - LENGTH(replace(v.key_skills,'\t','')) + 1) qtn
    FROM
        public.vacancies v

    WHERE 
        (name ILIKE '%data scientist%'
        OR name ILIKE '%data science%'
        OR name ILIKE '%исследователь данных%'
        OR (name LIKE '%ML%' AND name NOT ILIKE '%HTML%')
        OR name ILIKE '%machine learning%'
        OR name ILIKE '%машинн%обучен%')

        AND

        key_skills IS NOT NULL
''' 

In [50]:
# результат запроса
df22 = pd.read_sql_query(query_6_5, connection)
df22['qtn'].mean().round(2)

/usr/local/lib/python3.9/dist-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


6.41

In [51]:
# вариант подсчета с помощью Pandas
query_6_5_2 = f'''
    SELECT
        key_skills
    FROM
        public.vacancies v

    WHERE 
        (name ILIKE '%data scientist%'
        OR name ILIKE '%data science%'
        OR name ILIKE '%исследователь данных%'
        OR (name LIKE '%ML%' AND name NOT ILIKE '%HTML%')
        OR name ILIKE '%machine learning%'
        OR name ILIKE '%машинн%обучен%')

        AND

        key_skills IS NOT NULL
''' 

In [52]:
# результат запроса
df22_2 = pd.read_sql_query(query_6_5_2, connection)
df22_2['qtn_skills'] = df22_2['key_skills'].apply(lambda x: str(x).split('\t'))
df22_2['qtn_skills'] = df22_2['qtn_skills'].apply(lambda x: len(x))
df22_2['qtn_skills'].mean().round(2)

/usr/local/lib/python3.9/dist-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [53]:
# текст запроса
query_6_6 = f'''
    SELECT
        v.experience,
        AVG(CASE WHEN v.salary_from IS NULL THEN v.salary_to
            WHEN v.salary_to IS NULL THEN v.salary_from
            ELSE((v.salary_from + v.salary_to)/2)
        END) salary
    FROM
        public.vacancies v  
        
    WHERE 
        (name ILIKE '%data scientist%'
        OR name ILIKE '%data science%'
        OR name ILIKE '%исследователь данных%'
        OR (name LIKE '%ML%' AND name NOT ILIKE '%HTML%')
        OR name ILIKE '%machine learning%'
        OR name ILIKE '%машинн%обучен%')

        AND

        (salary_from IS NOT NULL OR salary_to IS NOT NULL)    

    GROUP BY 1   
'''

In [54]:
# результат запроса
df23 = pd.read_sql_query(query_6_6, connection)
df23

/usr/local/lib/python3.9/dist-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


experience         salary
0           Нет опыта   74642.857143
1  От 1 года до 3 лет  139674.750000
2       От 3 до 6 лет  243114.666667

# Выводы по предметному анализу
В базе имеется 1771 вакансия, которое имеет отношение к данным (3,6% от общего количества). При этом в 201 вакансии в качестве ключевого навыка указан SQL или postgres, в 351 вакансии в качестве ключевого навыка указан Python. В среднем в вакансиях, которые имеют отношение к данным, указывают 6.41 ключевых навыков. 
<br>Для начинающего дата-сайентиста подходит 51 вакансия (2,9% от всех вакансий имеющих отношение к данным).
<br>В среднем заработная плата в зависимости от опыта распределена следующим способом:
<br>Нет опыта - 74643
<br>От 1 года до 3 лет - 139675
<br>От 3 до 6 лет - 243115

# Общий вывод по проекту

In [55]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования